### PR TITLE
Deprecate REQUIRESSL privilege

### LINKED ISSUE #89: mysql_user idempotency issue with REQUIRESSL and ALL privileges
##### SUMMARY
There is an idempotency mismatch for certain privilege specifications. This probably happens because the names of certain privileges do not exactly match the names used by the database. See Ansible bug #29625 (which is not actually resolved for these cases).

Specific examples I found (for MariaDB 10.5)

    Specify privileges as "db.*:ALL,REQUIRESSL"
    Specify privileges as "db.*:ALL"

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

mysql_user

##### ANSIBLE VERSION

```paste below
ansible 2.9.16
  config file = /home/kkeane/Documents/dev/ansible/ansible.cfg
  configured module search path = [u'/home/kkeane/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug 13 2020, 02:51:10) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
```

##### CONFIGURATION

```paste below
ANSIBLE_NOCOWS(env: ANSIBLE_NOCOWS) = True
ANSIBLE_PIPELINING(/home/kkeane/Documents/dev/ansible/ansible.cfg) = True
DEFAULT_FORCE_HANDLERS(/home/kkeane/Documents/dev/ansible/ansible.cfg) = True
DEFAULT_FORKS(/home/kkeane/Documents/dev/ansible/ansible.cfg) = 10
DEFAULT_HOST_LIST(/home/kkeane/Documents/dev/ansible/ansible.cfg) = [u'/home/kkeane/Documents/dev/ansible/inventory/linux.yaml']
DEFAULT_MANAGED_STR(/home/kkeane/Documents/dev/ansible/ansible.cfg) = This file is managed by Ansible.%n
Do not edit this file. Changes will be reverted.%n
template: {file}
user: {uid}
host: {host}
DEFAULT_PRIVATE_KEY_FILE(/home/kkeane/Documents/dev/ansible/ansible.cfg) = /.{{ lookup('env','HOME') }}/.ssh/ansible_rsa
DEFAULT_REMOTE_USER(/home/kkeane/Documents/dev/ansible/ansible.cfg) = svc-ansible
DEFAULT_ROLES_PATH(/home/kkeane/Documents/dev/ansible/ansible.cfg) = [u'/home/kkeane/Documents/dev/ansible/roles', u'/home/kkeane/Documents/dev/ansible/download-roles', u'/usr/share/ansible/roles
DEFAULT_TIMEOUT(/home/kkeane/Documents/dev/ansible/ansible.cfg) = 300
DEFAULT_VAULT_PASSWORD_FILE(/home/kkeane/Documents/dev/ansible/ansible.cfg) = /home/kkeane/Documents/dev/ansible/vault-passwords
DISPLAY_SKIPPED_HOSTS(/home/kkeane/Documents/dev/ansible/ansible.cfg) = False
INVENTORY_ENABLED(/home/kkeane/Documents/dev/ansible/ansible.cfg) = [u'yaml', u'script', u'ini']
```

##### OS / ENVIRONMENT

Target OS: RHEL 8.3. MariaDB 10.5.8, installed from the MariaDB repository.

Also observed on some older versions of mysql.

Python 3.6.8 on the target.

##### STEPS TO REPRODUCE

(Note: derived from the actual playbook, but not tested verbatim)

```yaml
- name: Create user for this DB
  mysql_user:
    name: 'dbuser'
    host: 'fqdn.example.com'
    priv: "sampledb.*:ALL,REQUIRESSL"
    login_user: root
    login_host: localhost
    config_file: /root/.my.cnf
  become: True
```

##### EXPECTED RESULTS

On first run, the task may report changed.
On subsequent runs, the task should report unchanged.

After changing "REQUIRESSL" to "REQUIRE SSL", the same behavior should occur.

##### ACTUAL RESULTS

The task reports as changed on every run.
The task produces an error when changing REQUIRESSL to "REQUIRE SSL"

```paste below
Escalation succeeded
<redacted.example.com> (0, '\n{"changed": true, "user": "redacted_user", "msg": "Privileges updated", "invocation": {"module_args": {"config_file": "/root/.my.cnf", "name": "redacted_user", "login_user": "root", "login_host": "localhost", "host": "redacted.example.com", "priv": "redactedb.*:ALL,REQUIRESSL", "user": "redacted_user", "login_port": 3306, "encrypted": false,
"host_all": false, "state": "present", "append_privs": false, "check_implicit_admin": false, "update_password": "always", "connect_timeout": 30, "sql_log_bin": true, "login_password": null, "login_unix_socket": null, "password": null, "client_cert": null, "client_key": null, "ca_cert": null}}, "warnings": ["Module did not set no_log for update_password"]}\n', 'OpenSSH_7.4p1, OpenSSL 1.0.2k-fips  26 Jan 2017\r\ndebug1: Reading configuration data /home/kkeane/.ssh/config\r\ndebug2: checking match for \'user svc-ansible\' host redacted.example.com originally redacted.example.com\r\ndebug3: /home/kkeane/.ssh/config line 1: matched \'user "svc-ansible"\' \r\ndebug2: match found\r\ndebug1: /home/kkeane/.ssh/config line 71: Applying options for *.example.com\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 58: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 4 setting O_NONBLOCK\r\
ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 31162\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id:
2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\n')
changed: [redacted.example.com] => (item=redacted.example.com) => {
    "ansible_loop_var": "item",
    "changed": true,
    "invocation": {
        "module_args": {
            "append_privs": false,
            "ca_cert": null,
            "check_implicit_admin": false,
            "client_cert": null,
            "client_key": null,
            "config_file": "/root/.my.cnf",
            "connect_timeout": 30,
            "encrypted": false,
            "host": "redacted.example.com",
            "host_all": false,
            "login_host": "localhost",
            "login_password": null,
            "login_port": 3306,
            "login_unix_socket": null,
            "login_user": "root",
            "name": "redacted.example.com",
            "password": null,
            "priv": "redacteddb.*:ALL,REQUIRESSL",
            "sql_log_bin": true,
            "state": "present",
            "update_password": "always",
            "user": "redacted_user"
        }
    },
    "item": "redacted.sandiego.edu",
    "msg": "Privileges updated",
    "user": "redacted_user",
    "warnings": [
        "Module did not set no_log for update_password"
    ]
}
```
Changing the string "REQUIRESSL" to "REQUIRE SSL" results in:

```
failed: [redacted.example.com] (item=redacted.example.com) => {"ansible_loop_var": "item", "changed": false, "item": "redacted.example.com", "msg": "invalid privileges string: Invalid privileges specified: frozenset({'REQUIRE SSL'})"}
```

The underlying problem here is probably that the SHOW GRANTS and GRANT USAGE commands use the syntax "REQUIRE SSL" while the module only accepts "REQUIRESSL".

```
MariaDB [(none)]> show grants for redacteduser@'redacted.example.com';
+---------------------------------------------------------------------------------------------------------------------------------------------------+
| Grants for redacteduser@redacted.example.com                                                                                               |
+---------------------------------------------------------------------------------------------------------------------------------------------------+
| GRANT USAGE ON *.* TO `redacteduser`@`redacted.example.com` IDENTIFIED BY PASSWORD '*redacted' REQUIRE SSL |
| GRANT ALL PRIVILEGES ON `redacteddb`.* TO `redacted_user`@`redacted.example.com`                                                            |
+---------------------------------------------------------------------------------------------------------------------------------------------------+
```

Modified test case: changing the privileges to "redactedb.*:ALL" (without REQUIRESSL) also results in the task reporting "changed" on every run.

### PR DESCRIPTION
**Pull request by Jorge-Rodriguez**
##### SUMMARY

Parse the `priv` string to effectively turn the `REQUIRESSL` privilege into an alias for the equivalent setup in the `tls_requires` option.
Introduce a deprecation notice for the `REQUIRESSL` privilege.

Fixes #89 #121 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

mysql_user

### CODE REVIEW

**Comment by Andersson007** [comment]
```
ERROR: Found 4 pep8 issue(s) which need to be resolved:
ERROR: plugins/modules/mysql_user.py:975:40: E226: missing whitespace around arithmetic operator
ERROR: plugins/modules/mysql_user.py:978:40: E226: missing whitespace around arithmetic operator
ERROR: plugins/modules/mysql_user.py:978:61: E226: missing whitespace around arithmetic operator
ERROR: plugins/modules/mysql_user.py:978:128: E226: missing whitespace around arithmetic operator
```
maybe something else in CI

File: plugins/modules/mysql_user.py
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
        module.deprecate('The "REQUIRESSL" privilege is deprecated, use the "tls_requires" option instead.',
                          version='3.0.0', collection_name='community.mysql')
```

File: plugins/modules/mysql_user.py
Comment (Andersson007) [inline plugins/modules/mysql_user.py:970]:
```suggestion
def handle_requiressl_in_priv_string(module, priv, tls_requires):
```
@Jorge-Rodriguez i don't see where the function is invoked or I missed something

File: plugins/modules/mysql_user.py
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
    if "REQUIRESSL" in priv:
```
Maybe we could run this check somewhere outside and then invoke the function if "REQUIRESSL" in priv?
So we could avoid the invocation when it's not needed and avoid extra indentation inside the function.
What do you think?

File: plugins/modules/mysql_user.py
Comment (Jorge-Rodriguez) [reply plugins/modules/mysql_user.py:970]: Hah! Good one. It's invoked in my head, apparently. Let me fix that.

File: plugins/modules/mysql_user.py
Comment (Jorge-Rodriguez) [reply plugins/modules/mysql_user.py]: I purposefully wrote the check inside this function with the eye set on version 3.0.0

The idea is that when we drop support for `REQUIRESSL` we just need to remove this function and the key from the privileges frozen set.

But then again, we'd also be removing the function invocation which with your approach would be immediately underneath the check, will fix.

Comment (Andersson007) [reply plugins/modules/mysql_user.py]: Sounds good

**Comment by Jorge-Rodriguez** [comment]
Codecov is killing my OCD 😄

**Comment by Jorge-Rodriguez** [comment]
Ok, I was happily assuming that I wrote flawless code, but the fact that the integration tests were green when I removed all references to `REQUIRESSL` and forgot to invoke the new function would suggest that we weren't testing (properly) for `REQUIRESSL` so let's not merge this until I have proper tests in place.

**Comment by Andersson007** [comment]
> Ok, I was happily assuming that I wrote flawless code, but the fact that the integration tests were green when I removed all references to `REQUIRESSL` and forgot to invoke the new function would suggest that we weren't testing (properly) for `REQUIRESSL` so let's not merge this until I have proper tests in place.

Sure, let me know when it's ready for review

**Comment by Jorge-Rodriguez** [comment]
Ping @kkeane @Andersson007

**Comment by Andersson007** [comment]
@Jorge-Rodriguez should we review it now or better to wait until the CI is green?

**Comment by Jorge-Rodriguez** [comment]
@Andersson007 let's wait. I think I've spotted the issue with the failed CI and it's a systemic one on the module utils. I'm going to try to fix it and piggyback it on this PR

**Comment by Jorge-Rodriguez** [comment]
@Andersson007 ok, one changed my mind. This is ready for review.

I checked the issue and the fix is not as straightforward as I initially estimated and it warrants A PR of it's own.

I'm going to open a discussion about it based on [this comment](https://github.com/ansible-collections/community.mysql/issues/90#issuecomment-774628205)

File: tests/integration/targets/test_mysql_user/tasks/issue-121.yml
Comment (Andersson007) [inline tests/integration/targets/test_mysql_user/tasks/issue-121.yml:45]:
```suggestion
        priv: '*.*:SELECT,CREATE USER,REQUIRESSL,GRANT'
```
we could also check what is in user's privs in DB using `mysql_query` module to be more confident. Do you think if it's possible in this case?

**Comment by Andersson007** [comment]
@bmalynovytch your thoughts on this?

File: tests/integration/targets/test_mysql_user/tasks/issue-121.yml
Comment (Jorge-Rodriguez) [reply tests/integration/targets/test_mysql_user/tasks/issue-121.yml:45]: Yes, that's something we could do instead

**Review by bmalynovytch — APPROVED** [review]
LGTM

**Comment by Jorge-Rodriguez** [comment]
I will change the tests to what @Andersson007 suggested and that should make integration green before merging.

**Comment by Andersson007** [comment]
@Jorge-Rodriguez so, we will wait for the tests changes mentioned, right?

**Comment by Jorge-Rodriguez** [comment]
> @Jorge-Rodriguez so, we will wait for the tests changes mentioned, right?

@Andersson007 yes, I'll ping you again when it's ready.

**Comment by Jorge-Rodriguez** [comment]
Ping @Andersson007 
Ready

**Review by Andersson007 — APPROVED** [review]
LGTM
@Jorge-Rodriguez thanks for the great work!

**Comment by Andersson007** [comment]
So, we're ready to conduct our first major release soon. Maybe next week, say, on Monday:) will see

**Pull request merged**
